### DISCUSSION TITLE
"UNIQUE constraint failed: _collections.name" after upgrading from 0.22.27 to 0.23.12 against no existing pb_data

### DISCUSSION THREAD

**Discussion by seriousm4x** [pocketbase/pocketbase#6198] [Q&A] [answered]
Hi, I'm using pocketbase over at [upsnap](https://github.com/seriousm4x/UpSnap). I've just upgraded from pocketbase 0.22.27 to 0.23.12 following the [migration guide](https://pocketbase.io/v023upgrade/go/).

I get the following error when there is no pb_data:
```zsh
~/r/U/backend ❯❯❯ go run ./main.go serve --dir ./pb_data --http 0.0.0.0:8090         master ✱
Error: Failed to apply migration 1735377462_collections_snapshot.go: failed to save collection "users": constraint failed: UNIQUE constraint failed: _collections.name (2067)
```

How I migrated the codebase:
- I had an existing database compatible with 0.22.27
- Adjust my go code to be compatible with 0.23.12
- Deleted all go migration files
- `go run ./main.go --dir ./pb_data migrate collections`
- `go run ./main.go --dir ./pb_data migrate history-sync`
- The existing database migrated without issues

I tried to debug this and the migration works perfectly once I rename the `users`collection to something else in [line 699](https://github.com/seriousm4x/UpSnap/blob/master/backend/migrations/1735377462_collections_snapshot.go#L699).

It also works when pb_data existed from an older version before upgrading.

### How you can reproduce this:
- Download version [4.4.0](https://github.com/seriousm4x/UpSnap/releases/tag/4.4.0) of upsnap and run `./upsnap serve --dir=./pb_data`

**Comment by ganigeorgiev** [comment, marked as answer]
Changing the last parameter of [`ImportCollectionsByMarshaledJSON`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.12/core#BaseApp.ImportCollectionsByMarshaledJSON) to `true` in https://github.com/seriousm4x/UpSnap/blob/master/backend/migrations/1735377462_collections_snapshot.go#L1593 should fix it (_aka. it forces delete of missing/non-existing collections and fields, which was the default behavior in v0.22.x_).

The issue occur because in the older version you had deleted the default users collection as part of https://github.com/seriousm4x/UpSnap/blob/25373a66058f18e6fc5c62d92e0f1e3576874a79/backend/migrations/1674933029_deleted_users.go and then created your own new users collection in https://github.com/seriousm4x/UpSnap/blob/25373a66058f18e6fc5c62d92e0f1e3576874a79/backend/migrations/1677961980_created_users.go.

But since the migrations are deleted for the upgrade, the collections snapshot has no information about this change and your custom users collection is conflicting with the default PocketBase one that is created with the initial system migration in https://github.com/pocketbase/pocketbase/blob/master/migrations/1640988000_init.go#L81 (although they have the same name, they also have different ids - `_pb_users_auth_` vs `27do0wbcuyfmbmx`).

By changing the last parameter of `ImportCollectionsByMarshaledJSON` to `true` it instructs the import function to delete any collection or field that doesn't exist in the snapshot.

---

As a side-note, I noticed that you have/had your own installer but the default PocketBase v0.23+ init behavior may interrupt your desired flow because we now create a system superuser and automatically open a link in the browser to the PocketBase dashboard so that users can create their own account from there (the reason for this is explained in https://github.com/pocketbase/pocketbase/discussions/5814).

I'll consider with the next v0.24 release to add an option (maybe as part of the serve hook) to disable the default installer so that you can load your own custom one but a temporary workaround could be to create a superuser hook that will prevent the insertion of the system superuser:

```go
app.OnRecordCreate(core.CollectionNameSuperusers).BindFunc(func(e *core.RecordEvent) error {
	if e.Record.Email() == core.DefaultInstallerEmail {
		return errors.New("skip default PocketBase installer")
	}

	return e.Next()
})
```

**Reply by einarpersson** [reply]
@ganigeorgiev 

Hi, I'm finally doing this migration and have the same problem. It works when I run  against a copy of the production data, but when I start pocketbase without an existing db I get `Error: Failed to apply migration 1746532905_collections_snapshot.js: failed to save collection "users": constraint failed: UNIQUE constraint failed: _collections.name (2067)`.

In my case I think I renamed the original users collection and created another one. Here is an overview:

```json
      "name": "internalUsers",
      "id": "_pb_users_auth_",
      "indexes": [
        "CREATE UNIQUE INDEX `__pb_users_auth__username_idx` ON `internalUsers` (username COLLATE NOCASE)",
        "CREATE UNIQUE INDEX `__pb_users_auth__email_idx` ON `internalUsers` (`email`) WHERE `email` != ''",
        "CREATE UNIQUE INDEX `__pb_users_auth__tokenKey_idx` ON `internalUsers` (`tokenKey`)",
        "CREATE INDEX `__pb_users_auth__created_idx` ON \"internalUsers\" (`created`)"
      ],
```

```json
      "name": "users",
       "id": "ynf942m3txfbmuj",
      "indexes": [
        "CREATE UNIQUE INDEX `_ynf942m3txfbmuj_username_idx` ON `users` (username COLLATE NOCASE)",
        "CREATE UNIQUE INDEX `_ynf942m3txfbmuj_email_idx` ON `users` (`email`) WHERE `email` != ''",
        "CREATE UNIQUE INDEX `_ynf942m3txfbmuj_tokenKey_idx` ON `users` (`tokenKey`)"
      ],
```

However, changing `false` to `true` as suggested in the migrations file does not seem to help.

```js
  return app.importCollections(snapshot, true);
}, (app) => {
  return null;
```

I still get the same error. Do you have any idea on what is the problem? I am using js as you can see.

If not, is there some workaround, i.e. can I rename the collections in the migrations to resolve it somehow? I understand the problem on a high level _"custom users collection is conflicting with the default PocketBase one that is created with the initial system migration"_ but not enough to be able to resolve it. 

Grateful for any pointers / advice

**Reply by ganigeorgiev** [reply]
Setting the last parameter of `importCollections` to `true` should "clear" any collection missing from the json (_unless it was marked as "system"_) so make sure that you've saved the file.

Without access to the full collections snapshot I'm not sure what else could be the cause of the issue in your case.

**Reply by einarpersson** [reply]
Hmm I think I worked around it by renaming `users` to `externalUsers`. Toggling true/false did not make any difference. Perhaps the actual error is a bit different even though the error message is the same. Hopefully this works out.

**Reply by ganigeorgiev** [reply]
>  Toggling true/false did not make any difference.

We have automated tests that verify that its works correctly.

Start the executable with `--dev` and check the executed queries in the terminal.

There is definetely something strange going on in your setup that it is causing it to fail but without details about your running environment and a code sample with the failing migration I cannot help you.

**Comment by seriousm4x** [comment]
Thank you for figuring out the migration issue! That works perfectly for me. You are right, I deleted the users collection in the first place.

About the superuser problem: Your record hook skips the creation of the default superuser, but after that it is not possible to create a superuser via the api. "Only superusers can perform this action". Looking at the collection settings for _superusers in the dashboard, there are no api rules for superusers (which makes sense). So I guess my way of creating a superuser via custom installer is currently not possible with 0.23?

**Reply by ganigeorgiev** [reply]
It is kindof doable but you'll need to either do the same as we do in the [builtin installer](https://github.com/pocketbase/pocketbase/blob/master/apis/installer.go) or register a custom public route that will allow superusers creation only if there are no other superusers to emulate the old behavior. Something like (_I haven't tested it_):

```go
se.Router.POST("/api/upsnap/init-superuser", func(e *core.RequestEvent) error {
	superusersCollection, err := e.App.FindCollectionByNameOrId(core.CollectionNameSuperusers)
	if err != nil {
		return e.NotFoundError("Failed to retrieve superusers collection", err)
	}

	totalSuperusers, err := e.App.CountRecords(superusersCollection)
	if err != nil {
		return e.InternalServerError("Failed to retrieve superusers count", err)
	}

	if totalSuperusers > 0 {
		return e.BadRequestError("An initial superuser already exists", nil)
	}

	data := struct {
		Email    string `json:"email" form:"email"`
		Password string `json:"password" form:"password"`
	}{}
	err = e.BindBody(&data)
	if err != nil {
		return e.BadRequestError("Failed to read request data", err)
	}

	record := core.NewRecord(superusersCollection)
	record.SetEmail(data.Email)
	record.SetPassword(data.Password)
	err = e.App.Save(record)
	if err != nil {
		return e.BadRequestError("Failed to create initial superuser", err)
	}

	return apis.RecordAuthResponse(e, record, "", nil)
})
```

And replace in your client-side call `$pocketbase.admins.create({ ... })` with `$pocketbase.send("/api/upsnap/init-superuser", { method: "POST", body: { ... } })` and then eventually manually load the returned auth data into the store with `pb.authStore.save(authData.token, authData.record)`.

_Side-note: You may want to change the `_admins` references since with v0.23 is `_superusers`._

**Reply by seriousm4x** [reply]
Thanks a lot! Worked perfectly. Only thing I added was a `password_confirm` field. Thank you for your awesome support 💯

**Comment by fisocodes** [comment]
Hi! I'm facing the same issue. I downloaded the binary for my project. How can I solve this? Do I have to download the code and builld the binary with the sugested change? I tried to export and import the collections by JSON but got error with relations

**Reply by ganigeorgiev** [reply]
@fisocodes Please don't hijack discussions like this. The reason for the original error is explained in the comment above - https://github.com/pocketbase/pocketbase/discussions/6198#discussioncomment-11689322.

Just because you are getting similar error message doesn't mean that the cause of it is the same.

If you need help, open a new Q&A discussion but elaborate more on your use case and provide clear reproducible steps to trigger the error.
